ethier/sign.go

```diff
@@ -0,0 +1,141 @@
+package main
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/divergencetech/ethier/eth"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/spf13/cobra"
+)
+
+var signer *eth.Signer
+
+func init() {
+	var signCmd = &cobra.Command{
+		Use:   "sign",
+		Short: "Signs messages from stdin using an ECDSA signer.",
+	}
+
+	rootCmd.AddCommand(signCmd)
+	signCmd.PersistentFlags().String("mnemonic", "", "Path to file containing the mnemonic, password and account number (newline separated) for key derivation")
+
+	var signAddrCmd = &cobra.Command{
+		Use:     "addresses",
+		Short:   "Signs addresses from stdin using an ECDSA signer.",
+		PreRunE: getSigner,
+		RunE:    signAddresses,
+	}
+
+	signCmd.AddCommand(signAddrCmd)
+}
+
+type SignedAddress struct {
+	Address   string `json:"address"`
+	Signature string `json:"signature"`
+}
+
+func getSignerFromMnemonic(filepath string) (*eth.Signer, error) {
+	dat, err := os.ReadFile(filepath)
+	if err != nil {
+		return nil, fmt.Errorf("Read mnemonic: %v", err)
+	}
+
+	lines := strings.Split(string(dat), "\n")
+
+	mnemonic := lines[0]
+	var password string = ""
+	var account uint = 0
+
+	if len(lines) > 1 {
+		password = lines[1]
+	}
+
+	if len(lines) > 2 {
+		if lines[2] != "" {
+			acc, err := strconv.Atoi(lines[2])
+			if err != nil {
+				return nil, fmt.Errorf("Converting derivation account: %v", err)
+			}
+			account = uint(acc)
+		}
+	}
+
+	signer, err := eth.DefaultHDPathPrefix.SignerFromSeedPhrase(
+		mnemonic, password, account,
+	)
+
+	if err != nil {
+		return nil, fmt.Errorf("Generate signer: %v", err)
+	}
+	return signer, nil
+}
+
+func getSigner(cmd *cobra.Command, args []string) (retErr error) {
+	mnemonicPath, err := cmd.Flags().GetString("mnemonic")
+	if err != nil {
+		log.Fatalf("Get mnemonics flag: %v", err)
+	}
+
+	if mnemonicPath != "" {
+		signer, err = getSignerFromMnemonic(mnemonicPath)
+		if err != nil {
+			log.Fatalf("Generate signer from mnemonic: %v", err)
+		}
+	} else {
+		signer, err = eth.NewSigner(256)
+		if err != nil {
+			log.Fatalf("Generate new signer: %v", err)
+		}
+	}
+
+	return nil
+}
```

Comment:
Given that we'll push the user-provided signer to a later date, I think the `signer, err := eth.NewSigner()` bit can just be merged into `signAddresses()`.

As a side note, if we were keeping this function I think it should've been called directly by `signAddresses()` instead of being used as a `PreRun` on the Cobra command. The indirection of the `signer` via a global variable has undesirable effects of (a) putting a sensitive value in a global scope; and (b) creating a non-linear control flow that confuses readers of the code.

Reply:
I agree, there are only disadvantages in breaking the linear code flow here. I reverted the introduction of reading mnemoics, and merged `signer, err := eth.NewSigner()` again with `signAddresses`.